### ISSUE TITLE
`Target` does not enforce implied invariants

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#1828]
`Target` has an associated const `MAX` which implies in invariant that self is not greater than `MAX`. 

One can create target/work types using constructors `from_be/le_bytes` and invalidate these invariants. 

Should we enforce these invariants?

**Comment by Kixunil** [comment]
Hmm maybe. `from_${endian}_bytes` will need to return `Result` then which is a bit surprising but not wrong.

**Comment by Kixunil** [comment]
> Should we enforce that?

I think if we're enforcing bounds we should enforce these too.

> why do we support encoding these in any format except the 32-bit compact-bits encoding?

API completeness, debugging... Seems useful enough.

**Comment by mpbagot** [comment]
`Work` no longer has any min/max constants, nor does it have a restriction like encoding in a compact form. Should this be revised down to just cover `Target`?

Also, if this needs to enforce restrictions on `MAX_ATTAINABLE_*` (based on the required network) instead of just `MAX` in `Target`, then should the existing `from_*_bytes` be renamed to `from_*_bytes_unchecked`, and new `from_*_bytes` take a `Params`?

**Comment by tcharding** [comment]
We cannot actually do this without intorducing `Network` because since writing this issue we added network specific consts.

To close this I think we should just document that `Target` does not enforce the invariant that it is smaller that the max value (for the current network).

**Comment by tcharding** [comment]
Requires conceptACK by @apoelstra before implementation to save doing unnecessary work or review.

**Comment by apoelstra** [comment]
Yep, concept ACK. Since the "maximum" is network-dependent it no longer makes sense as a type-level invariant.
